### PR TITLE
Check total field limit at parse time

### PR DESCRIPTION
**Pull request by ywelsch**
If dynamic mapping updates are enabled, having a crazy number of fields in a document can generate a very large dynamic mapping update that in turn can make a node go OOM even before the mapping update is validated. This commit adds a pre-flight check on the total number of fields allowed in a mapping at document parse time, while the (potentially large) dynamic mapping update is being built.

Relates #73460

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/index/mapper/NestedObjectMapperTests.java [hunk 1]
```diff
@@ -625,9 +625,12 @@ public void testParentObjectMapperAreNested() throws Exception {
     }
 
     public void testLimitNestedDocsDefaultSettings() throws Exception {
-        Settings settings = Settings.builder().build();
+        Settings settings = Settings.builder()
+            // make sure we're not hitting the total field limit
+            .put(MapperService.INDEX_MAPPING_TOTAL_FIELDS_LIMIT_SETTING.getKey(), Long.MAX_VALUE)
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/index/mapper/NestedObjectMapperTests.java:630]: out of curiosity: why was this change required?

File: server/src/internalClusterTest/java/org/elasticsearch/index/mapper/DynamicMappingIT.java [hunk 2]
```diff
@@ -154,10 +157,48 @@ public void onFailure(String source, Exception e) {
...
+                    masterBlockedLatch.countDown();
+                    indexingCompletedLatch.await();
+                    return currentState;
+                }
+
+                @Override
+                public void onFailure(String source, Exception e) {
+                    throw new AssertionError("unexpected", e);
+                }
+            });
```
Comment (javanna) [inline server/src/internalClusterTest/java/org/elasticsearch/index/mapper/DynamicMappingIT.java:192]: could you expand, and maybe leave a comment, on why this cluster state update is needed?

File: server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java [hunk 3]
```diff
@@ -207,7 +207,11 @@ void checkLimits(IndexSettings settings) {
     }
 
     private void checkFieldLimit(long limit) {
-        if (fieldMappers.size() + objectMappers.size() - mapping.getSortedMetadataMappers().length > limit) {
+        checkFieldLimit(limit, 0);
+    }
+
+    public void checkFieldLimit(long limit, int additionalFieldsToAdd) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:213]: nit: I find it a bit odd that we take an int but the args are always 0 or 1. maybe we should have two methods "checkFieldLimit" and "canAddNewField" or something along those lines? Also I would prefer that this method is package private. MappingLookup is not really meant to expose these to the outside world (at least outside its package).

**Review by javanna — APPROVED** [review]
Thanks @ywelsch I left a couple of comments but this LGTM

File: server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java (hunk 3 above)
Comment (markharwood) [reply server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:213]: >but the args are always 0 or 1.

The ParseContext class passes `dynamicMappers.size() + 1` not 1.

File: server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java [hunk 4]
```diff
@@ -207,7 +207,11 @@ void checkLimits(IndexSettings settings) {
...
+        if (fieldMappers.size() + objectMappers.size() + additionalFieldsToAdd - mapping.getSortedMetadataMappers().length > limit) {
             throw new IllegalArgumentException("Limit of total fields [" + limit + "] has been exceeded");
```
Comment (markharwood) [inline server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:215]: I wonder  if we can be more helpful with our messages here. 
There's 2 scenarios:
1) "The straw that broke the camel's back" - i.e.  there was steady growth towards the field limit over many docs.
2) A rogue document.

For 1) maybe we  could log warnings as we get close to hitting the limit - kind of like disk space threshold warnings. This would allow users the ability to react/redesign prior  to disaster.
For 2) maybe we should quantify how many _new_ fields a single document was attempting to add.

**Review by markharwood — APPROVED** [review]
Added a comment on  messages/warnings but otherwise LGTM

File: server/src/test/java/org/elasticsearch/index/mapper/NestedObjectMapperTests.java (hunk 1 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/index/mapper/NestedObjectMapperTests.java]: Thanks for this question: Looking into this again made me realize that my code had a bug. There can be multiple dynamic updates for the same field (which is only merged later), and this test runs exactly into that scenario (where there are nested docs that all have the same field, and there are more than 1000 of them). This wrongly triggers the field count check (as all fields are the same). Except for interleaving the process of merging different field definitions with the parsing code (i.e. with `DocumentParser.createDynamicUpdate`), we need to relax the check a bit here and can only count the number of different field names. This is fixed in 12de303816d

File: server/src/internalClusterTest/java/org/elasticsearch/index/mapper/DynamicMappingIT.java (hunk 2 above)
Comment (ywelsch) [reply server/src/internalClusterTest/java/org/elasticsearch/index/mapper/DynamicMappingIT.java:192]: This uses the same approach as the other test (that we modified), which ensures that the validation is happening on the data node before the mapping is processed on the master.

File: server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java (hunk 3 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java]: I've made the method package-private (see 8566c61). As Mark said, we're not just passing 0 and 1.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java]: right, my eyes tricked me there :)

File: server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java (hunk 4 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java]: I've expanded the message to cover scenario 2 in 8bda5d8.
I'm not sure about how common scenario 1 is. I haven't seen cases where this limit was slowly reached. I'm doubtful if logging would help, perhaps some kind of monitoring with Kibana UI integration would help. Or part of a more general "cluster health" API. Not something I would like to address as part of this PR.

File: server/src/test/java/org/elasticsearch/index/mapper/NestedObjectMapperTests.java (hunk 1 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/index/mapper/NestedObjectMapperTests.java]: Urgs, had to follow this up with another commit: cd79731

File: server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java [hunk 5]
```diff
@@ -427,6 +428,10 @@ public void seqID(SeqNoFieldMapper.SequenceIDFields seqID) {
 
         @Override
         public void addDynamicMapper(Mapper mapper) {
+            // eagerly check field name limit here to avoid OOM errors
+            if (mappingLookup.getMapper(mapper.name()) == null && newFieldsSeen.add(mapper.name())) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java:432]: tricky bits: what happens when we dynamically add an object field? and what is the scenario where the mapper is not found (meaning it's a new one?) but is already in the set?

**Review by javanna — COMMENTED** [review]
I left one question: small change but not super straight-forward, thanks for taking the time.

File: server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java (hunk 5 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java:432]: - object fields are not treated specially in this logic (it's counted as a proper field, similar as it's done in the field limit check)
- the scenario where the mapper is not found but is already in the set covers scenarios like the one found in the `testLimitNestedDocsDefaultSettings` test:
```
"foo" : [ { "bar" : 1 }, { "bar" : 2 }, { "bar" : 3} ]
```
If the "bar" field has not been mapped prior to receiving this document, parsing this doc results in 3 separate calls to `addDynamicMapper`. Ultimately these are merged together in `DocumentParser.createDynamicUpdate`, so should only count as 1 (and will only count as 1 for the fields limit check that is run on the merged mapping). If we didn't have this logic, we would now start failing the fields limit check at parse time because we would count these as separate fields even though it's just one.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java:432]: makes sense, I had forgotten that we have the same mapper multiple times, and we merge those to prevent conflicts between them. 

My question around object fields revolved around the fact that mappingLookup#getMapper does not seem expose object mappers, so I was wondering about the behaviour when an object with the provided name is defined. I guess that the check will always return null hence we may count objects that are already mapped as new? I may be missing something but maybe it is worth adding a test for this scenario to verify what happens?

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java:432]: oh well maybe that never happens because we only call addDynamicMapper for fields that are not known... still may make sense to verify that we are doing the right thing.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java:432]: I guess that my next question is why we need the getMapper call. When are we calling addDynamicMapper providing a mapper that mapping lookup is already aware of?

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java:432]: > When are we calling addDynamicMapper providing a mapper that mapping lookup is already aware of?

I wasn't sure how much we can rely on that. For ObjectMapper I felt that this would not be an issue, but for other arbitrary FieldMapper (which can be added by plugins etc.) I was wondering if the code should have an additional guard against that (e.g. ConstantKeywordFieldMapper directly calls this, and even though the guard is not necessary for this class, it's unclear to me everyone would stick to this contract). If we have for example a new enum field type that would allow values to be populated dynamically, a dynamic mapper would possibly want to add a new enum value to existing enum field (i.e. dynamic mapping update to the existing field).

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java:432]: oh boy! you are right that we can't rely on it. I find it pretty scary and surprising that field mappers can add dynamic mappers themselves, I was only looking at the ordinary flow where this method is called when parsing a document because the field is not already mapped. Could you add a small comment that explains why we have that if, and possibly a test that fails if we remove it? I think that you could use constant keyword in your test to show how an already mapped field can be among the dynamic mappers yet it should not be counted. Thanks for going through this with me :)

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java]: I've pushed a commit (e596aea) that adds comments to the code, as well as a test for constant_keyword (that fails when you remove the extra check whether the field already exists). I've also expanded the check to cover object mappers. Even though I can't construct such a test case, I would like to be extra careful in this place.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java]: Thanks! if you really wanted to test also the object mapper scenario I think you could define a custom field that as part of its parseCreateField adds a dynamic mapper for an object field. It is not required, it sounds like a crazy usecase and stuff we may look at in the future and wonder why we tested it.

**Review by javanna — APPROVED** [review]
LGTM thanks @ywelsch !

File: server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java (hunk 5 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java]: yes, I find it more confusing to add such a synthetic test.
